src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.OSX.cs

```diff
@@ -1,22 +1,183 @@
 ﻿// Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+using Microsoft.Win32.SafeHandles;
+using System.Diagnostics;
+using System.Threading;
+
+using CFStringRef = System.IntPtr;
+using CFRunLoopRef = System.IntPtr;
+
 namespace System.Net.NetworkInformation
 {
     // OSX implementation of NetworkChange
-    // TODO: This can be implemented using the SystemConfiguration framework.
-    // See <SystemConfiguration/SystemConfiguration.h> and its documentation.
+    // See <SystemConfiguration/SystemConfiguration.h> and its documentation, as well as
+    // the documentation for CFRunLoop for more information on the components involved.
     public class NetworkChange
     {
-        static public event NetworkAddressChangedEventHandler NetworkAddressChanged
+        private static object s_lockObj = new object();
+
+        // The list of current subscribers.
+        private static NetworkAddressChangedEventHandler s_addressChangedSubscribers;
+
+        // The dynamic store. We listen to changes in the IPv4 and IPv6 address keys.
+        // When those keys change, our callback below is called (OnAddressChanged).
+        private static SafeCreateHandle s_dynamicStoreRef;
+
+        // The RunLoop source, created over the above SCDynamicStore.
+        private static SafeCreateHandle s_runLoopSource;
+
+        // Listener thread that adds the RunLoopSource to its RunLoop.
+        private static Thread s_runLoopThread;
+
+        // The listener thread's CFRunLoop.
+        private static CFRunLoopRef s_runLoop;
+
+        // Use an event to try to prevent StartRaisingEvents from returning before the
+        // RunLoop actually begins. This will mitigate a race condition where the watcher
+        // thread hasn't completed initialization and stop is called before the RunLoop even starts.
```

Comment:
I think it's correct, if you're referring to the second-last word. Perhaps I should reword it to "before the run loop has even started". Or just remove "even" :smile: